borg/archive.py

```diff
@@ -298,7 +298,36 @@ def add_file_chunks(chunks):
         cache.rollback()
         return stats
 
-    def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sparse=False):
+    def get_link_chunks(self, path, items):
+        """
+        path, items -> (can_link, matched_path, chunk_ids, link_target)
+
+        In other words, when extracting subtrees of an archive this determines if a link is possible
+        within the subtree, or a link needs to "be broken", i.e. turned into a regular file.
+        """
+        while True:
+            chunks, linked_to = items[path]
+            if linked_to in items:
+                path = linked_to
+                continue
+            if chunks is not None:
+                return False, path, chunks, None
+            if os.path.exists(linked_to):
+                return True, path, None, linked_to
+
+    def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sparse=False, items=None,
+                     original_path=None):
+        """
+        Extract archive item.
+
+        :param item: the item to extract
+        :param restore_attrs: restor file attributes
+        :param dry_run: do not write any data
+        :param stdout: write extracted data to stdout
+        :param sparse: recreate sparse files (only approximate, no exact hole reproduction)
```

Comment:
in fact it is even a "sparsify" of all files that have all-zero chunks, no matter whether they were holes a archive creation time or not.